packages/labs/src/controllers/context.ts

```diff
@@ -0,0 +1,324 @@
+import {UpdatingController} from './updating-controller.js';
+export * from './updating-controller.js';
+import {notEqual, UpdatingElement} from 'updating-element';
+import {
+  AttributePart,
+  ATTRIBUTE_PART,
+  directive,
+  Directive,
+  nothing,
+  Part,
+  PROPERTY_PART,
+  DirectiveClass,
+  DirectiveParameters,
+} from 'lit-html';
+
+// TODO(sorvell): lit-html should export this?
+declare type DirectiveResult<C extends DirectiveClass = DirectiveClass> = {
+  _$litDirective$: C;
+  values: DirectiveParameters<C>;
+};
+
+const connectContextEvent = 'connect-context';
+
+const providerMap: WeakMap<Node, Provider[]> = new WeakMap();
+
+export type ConsumerEvent = CustomEvent<Consumer[]>;
+
+export const EVENT_STRATEGY = 1;
+export const ASCEND_STRATEGY = 2;
+
+const DEFAULT_STRATEGY = EVENT_STRATEGY;
+
+// Note, this currently wouldn't strictly have to be an UpdatingController.
+export class Provider extends UpdatingController {
+  strategy = DEFAULT_STRATEGY;
+  consumers: Set<Consumer> = new Set();
+  key = null;
+
+  // Note, these are not private so they can be used in mixins.
+  // @internal
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  _value: any = null;
+  // @internal
+  _directive?: () => DirectiveResult;
+
+  constructor(host: UpdatingElement, value?: unknown) {
+    super(host);
+    this.value = value;
+  }
+
+  provideAll() {
+    this.addConnectListener(this.host);
+  }
+
+  provide() {
+    if (this._directive === undefined) {
+      // eslint-disable-next-line @typescript-eslint/no-this-alias
+      const provider = this;
+      const ProviderDirective = class extends Directive {
```

Comment:
It's not a new class per `provide()` call, but per instance of `Provider`, which I think is the design (so you don't have to pass `provider` to the directive invocation.

We could flip it around, so the user has to pass their `Provider` instance to a generic `provide` directive, rather than using a specific directive produced by the `Provider`; is that what you're suggesting?

Reply:
You can make a generic `provide(provider)` directive, then still have a method here that uses it:

```ts
const providerDirective = directive(class extends Directive { ... });

class Provider {
  provide() {
    return provideDirective(this);
  }
}
```

```ts
html`<div ${provider.provide()}`></div>
```